go/pkg/command/command.go

```diff
@@ -0,0 +1,273 @@
+	ExecRoot string
+
+	// The working directory, relative to the exec root, for the command to run
+	// in. It must be a directory which exists in the input tree. If it is left
+	// empty, then the action is run in the exec root.
+	WorkingDir string
+
+	// The command inputs.
+	InputSpec *InputSpec
+
+	// The command output files.
+	OutputFiles []string
+
+	// The command output directories.
+	// The files and directories will likely be merged into a single Outputs field in the future.
+	OutputDirs []string
+
+	// Optional duration to wait for command execution before timing out.
+	Timeout time.Duration
+
+	// The platform to use for the remote execution.
+	Platform map[string]string
+}
+
+func marshallMap(m map[string]string, buf *[]byte) {
+	var pkeys []string
+	for k := range m {
+		pkeys = append(pkeys, k)
+	}
+	sort.Strings(pkeys)
+	for _, k := range pkeys {
+		*buf = append(*buf, []byte(k)...)
+		*buf = append(*buf, []byte(m[k])...)
+	}
+}
+
+func marshallSlice(s []string, buf *[]byte) {
+	for _, i := range s {
+		*buf = append(*buf, []byte(i)...)
+	}
+}
+
+func marshallSortedSlice(s []string, buf *[]byte) {
+	ss := make([]string, len(s))
+	copy(ss, s)
+	sort.Strings(ss)
+	marshallSlice(ss, buf)
+}
+
+// Validate checks whether all required command fields have been specified.
+func (c *Command) Validate() error {
+	if c.Args == nil {
+		return errors.New("missing command arguments")
+	}
+	if c.ExecRoot == "" {
+		return errors.New("missing command exec root")
+	}
+	if c.InputSpec == nil {
+		return errors.New("missing command input spec")
+	}
+	if c.Identifiers == nil {
+		return errors.New("missing command identifiers")
+	}
+	return nil
+}
+
+// Generates a stable id for the command.
+func (c *Command) stableId() string {
+	var buf []byte
+	marshallSlice(c.Args, &buf)
+	buf = append(buf, []byte(c.ExecRoot)...)
+	buf = append(buf, []byte(c.WorkingDir)...)
+	marshallSortedSlice(c.OutputFiles, &buf)
+	marshallSortedSlice(c.OutputDirs, &buf)
+	buf = append(buf, []byte(c.Timeout.String())...)
+	marshallMap(c.Platform, &buf)
+	if c.InputSpec != nil {
+		marshallMap(c.InputSpec.EnvironmentVariables, &buf)
+		marshallSortedSlice(c.InputSpec.Inputs, &buf)
+		inputExclusions := make([]*InputExclusion, len(c.InputSpec.InputExclusions))
+		copy(inputExclusions, c.InputSpec.InputExclusions)
+		sort.Slice(inputExclusions, func(i, j int) bool {
+			e1 := inputExclusions[i]
+			e2 := inputExclusions[j]
+			return e1.Regex > e2.Regex || e1.Regex == e2.Regex && e1.Kind > e2.Kind
+		})
+		for _, e := range inputExclusions {
+			buf = append(buf, []byte(e.Regex)...)
+			buf = append(buf, []byte(e.Kind.String())...)
+		}
+	}
+	sha256Arr := sha256.Sum256(buf)
+	return hex.EncodeToString(sha256Arr[:])[:8]
+}
+
+// FillDefaultFieldValues initializes valid default values to inner Command fields.
+// This function should be called on every new Command object before use.
+func (c *Command) FillDefaultFieldValues() {
+	if c == nil {
```

Comment:
You don't check for this in other functions?

Reply:
I do now :-)